File: nodes/self-healing-error/self-healing-error.process.py
async def process(inputs, settings, config, nodeConfig):
    """
    Process function for the Self Healing Error node.
    Fails initially but succeeds after specified number of attempts.
    """
    value = inputs.get("value")
    fail_count = int(inputs.get("failCount", 1))
    error_message = inputs.get("errorMessage", f"Self-healing error: failing {fail_count} time(s) before success")
    
    # Get or initialize the execution count for this node instance
    if "_nodeState" not in config:
        config["_nodeState"] = {}
    
    node_id = config.get("_nodeId", "self-healing-error")
    if node_id not in config["_nodeState"]:
        config["_nodeState"][node_id] = {"executionCount": 0}
    
    state = config["_nodeState"][node_id]
    state["executionCount"] += 1
    
    # Raise error if we haven't reached the success threshold yet
    if state["executionCount"] <= fail_count:
        class SelfHealingError(Exception):
            def __init__(self, message, attempt, total_attempts):
                super().__init__(message)
                self.attempt = attempt
                self.total_attempts = total_attempts
                self.name = "SelfHealingError"
        
        error_msg = f"{error_message} (attempt {state['executionCount']}/{fail_count + 1})"
        raise SelfHealingError(error_msg, state["executionCount"], fail_count + 1)
    
    # Success! Return the value
    return {"result": value} 

File: sdks/nodejs/nodes/self-healing-error/self-healing-error.process.js
export default async ({inputs, settings, config}) => {

  const value = inputs.value;
  const failCount = parseInt(inputs.failCount) || 1;
  const errorMessage = inputs.errorMessage || `Self-healing error: failing ${failCount} time(s) before success`;
  
  // Get or initialize the execution count for this node instance
  if (!config._nodeState) {
    config._nodeState = {};
  }
  
  const nodeId = config._nodeId || 'self-healing-error';
  if (!config._nodeState[nodeId]) {
    config._nodeState[nodeId] = { executionCount: 0 };
  }
  
  const state = config._nodeState[nodeId];
  state.executionCount++;
  
  // Throw error if we haven't reached the success threshold yet
  if (state.executionCount <= failCount) {
    const error = new Error(`${errorMessage} (attempt ${state.executionCount}/${failCount + 1})`);
    error.name = 'SelfHealingError';
    error.attempt = state.executionCount;
    error.totalAttempts = failCount + 1;
    throw error;
  }
  
  // Success! Return the value
  return { result: value };
}; 